### ISSUE TITLE
User can check which models are up-to-date from config_log

### ISSUE THREAD

**Issue by seth127** [metrumresearchgroup/rbabylon#30]
# Summary
The tibble output from `config_log()` contains the following:
* path to model file (control stream)
* md5 hash of model file
* path to data file
* md5 hash of data file

The user should be able to easily check if any of those paths refer to a file that does _not_ match the md5 hash stored in the log tibble. This would indicate that the model is "out-of-date" and should be refreshed in the appropriate way (i.e. probably re-run).

**Comment by dpastoor** [comment]
To your question on the pr really like the idea of providing a couple helpers afternoon the hash comparations so people don't have to write those steps out

**Comment by dpastoor** [comment]
As of now - helper code supplied in vignette while we consider cases of how to use.

**Comment by seth127** [comment]
@kylebaron proposed that this is just baked into the `config_log()` function. As in, it already gives you everything you would need to calculate it, let's just calculate it and then we can have the `model_md5_match` and `data_md5_match` columns just be included automatically. 

I like that idea to be honest. It's simple and it accomplishes the goal. The only downside is possible wasted computation time, but that's likely going to be very small. We could even have an arguemnt to turn this off if we're really concerned about that but I'm not really. Any thoughts @dpastoor ?

**Comment by copernican** [comment]
The path to the model file is stored in `bbi_config.json` as `model_path`. 

The path to the data file can be constructed from the path to the JSON, and the `data_path` field from the JSON, as in the "Using the based_on field" vignette. The directory containing the JSON also appears to be stored in the JSON as `output_dir`, so we can take it from there as well.

We have access to all of this information in `config_log_impl()`, so the implementation should be simple.

1. Construct the paths to the model and data files.
2. Compute the MD5 sums of those files.
3. Compare those MD5 sums to the respective values stored in the JSON.
4. Add two new variables holding the results of those comparisons to the returned tibble.

While descriptive, the proposed names of the new variables (`model_md5_match` and `data_md5_match`) "hard-code" the comparison method (MD5 sums). Also, including words like "is" or "has" in the names of logical variables is common to convey meaning. Would we consider names like `data_is_current`, `data_has_changed`, `model_is_stale`, or even `data_matches`?

Whatever names we choose, the resulting tibble will be returned to `config_log()`, which will return it to the user. Does that sound like a reasonable plan?

**Comment by seth127** [comment]
That all sounds reasonable. The column names can change to whatever you think is most appropriate.

**Comment by copernican** [comment]
Should `add_config()` also add these new columns to its argument `.log_df`?

**Comment by seth127** [comment]
yes, these should be added to the `config_log_impl` function so that anything that calls it will have them. 

Also, you may have to modify something in `classes.R` or the corresponding constants in `aaa.R` for the added fields, although I'm not 100% sure on that.

**Comment by copernican** [comment]
We actually cannot directly use the `model_path` or `output_dir` fields, because they may refer to paths that do not exist, e.g., paths within the home directory of the user who ran the model. In the case of the model used for testing, we have the structure

```
.
└── testthat
    ├── data
    │   └── acop.csv
    └── model-examples
        ├── 1
        │   ├── 1.ctl
        │   └── bbi_config.json
        ├── 1.ctl
        └── 1.yaml
```

Getting the path to the data file is straightforward (`data_path` is relative to the output directory), and we have the path to the output directory because it contains the JSON. There are outstanding questions about the path to the model file, though.

1. Why are there two versions of `1.ctl`? They aren't identical:

    ```sh
    ➜  model-examples diff 1.ctl 1/1.ctl
    39c39
    < $COV
    ---
    > $COV
    \ No newline at end of file
    ```

2. Which version of `1.ctl` is the actual model file? If it's the one in the same directory as `bbi_config.json`, i.e., the output directory, then the immediate problem is somewhat easier, as we know the path to the JSON. If it's the one in the parent directory, then it may be trickier, unless it will always be the case that the model file is in the parent directory of the output directory.

**Comment by seth127** [comment]
Yes, I remember dealing with this when we were coding this up for the vignette. You can see the way we do it in the [vignette here](https://metrumresearchgroup.github.io/rbabylon/articles/using-based-on.html#checking-if-models-are-up-to-date-with-config_log), specifically:
```
log_df <- config_log()
norm_data_paths <- fs::path_norm(file.path(log_df$absolute_model_path, log_df$data_path))
norm_model_paths <- get_model_path(log_df)
```

To answer your questions explicitly:

> 1. Why are there two versions of 1.ctl? They aren't identical...

I don't remember the exact reason but I think it has to do with `bbi` and maybe grid submission. @dpastoor or @shairozan probably have a better answer but it's not necessary for this actual function because...

> 2. Which version of 1.ctl is the actual model file? 

The file which has its md5 stored in `bbi_config.json` under `model_md5` is the original model file (in the parent dir of the output dir). I am _fairly sure_ it will always be in the parent dir of the output dir, but I used `get_model_path()` above because there is theoretically a way for a user to store the model file in a different place but, if they did, it would need to be specified in the YAML which `get_model_path()` would retrieve.

**Comment by copernican** [comment]
As discussed over Slack, one meaning (currently) of `absolute_model_path` is the path to the model file without file extension, e.g., `.ctl`. Thus, we can construct the path to the model file from `absolute_model_path` and  the `model_extension` field from the JSON.

**Comment by copernican** [comment]
It does not appear to be the case that `model_md5` corresponds to the model file in the parent directory of the output directory. In this case, it is MD5 sum of the model file in the output directory.

```sh
➜  model-examples  md5sum 1.ctl 1/1.ctl
6b930119c4224ba077091b47959b0604  1.ctl
731923458236cc008c3adafa2f0877a7  1/1.ctl
➜  model-examples grep model_md5 1/bbi_config.json
    "model_md5": "731923458236cc008c3adafa2f0877a7",
```

**Comment by copernican** [comment]
closed by #148

**Comment by seth127** [comment]
updating description for validation. Original description:

# Summary
The tibble output from `config_log()` contains the following:
* path to model file (control stream)
* md5 hash of model file
* path to data file
* md5 hash of data file

The user should be able to easily check if any of those paths refer to a file that does _not_ match the md5 hash stored in the log tibble. This would indicate that the model is "out-of-date" and should be refreshed in the appropriate way (i.e. probably re-run).
